### ISSUE TITLE
Allow fat-arrow on a separate line

### ISSUE THREAD

**Issue by LukaHorvat** [atom-haskell/language-haskell#81]
These two declarations should be colored the same

![](http://i.imgur.com/1uJf6Oc.png)

Would it be possible to fix that?

**Comment by lierdakil** [comment]
Yes and no.

As for 'no' part, TextMate's grammar model (used by Atom) it's impossible to detect `=>` on the next line. See https://github.com/atom-haskell/language-haskell/issues/57 for more information.

As for 'yes' part, I just pushed v1.9.0 which plain and simply gives up on trying to detect constraints in type signatures, and just highlights prelude typeclasses as typeclasses, so that would actually make your definitions the same color, although not as nicely as one might hope.
![image](https://cloud.githubusercontent.com/assets/7275622/17486346/25fff556-5d99-11e6-8958-b8d0b22cdc29.png)

**Comment by LukaHorvat** [comment]
Well, that looks nice. I was wondering, would it be possible to highlight
identifiers that are in the current scope, differently? That sounds like a
pretty useful feature.

I guess it would just consist of updating your "known classes" dynamically.
I assume you already do a similar thing with the autocomplete lists.

**Comment by lierdakil** [comment]
No, no. Atom's grammar model doesn't support actual *parsing*,  only tokenizing, so no dynamic updates or anything, I'm afraid.

**Comment by LukaHorvat** [comment]
Hmm, that's a shame. I'm closing this issue.

I would like to hear your opinion about something like this though: https://atom.io/packages/language-javascript-semantic

**Comment by lierdakil** [comment]
`language-javascript-semantic` hacks into Atom's internal grammar representation and runs an external tokenizer from there (acorn.js in this case). Then it uses token string hash to assign color to variable. It's not context-aware or anything like that. Source is still parsed line-by-line, with no backtracking. It's a neat hack, but doesn't really apply to Haskell.

That said, something like this could be possible with display layers in post, although that would require to parse whole source file (which isn't exactly fast, so would have to be done asynchronously). But that's an issue entirely separate from actual grammar definition.

**Comment by LukaHorvat** [comment]
Well yeah, but isn't this enough? When I say "identifiers in scope" I just mean things that are available from the imports. You must have a list of those somewhere to provide autocompletions. Couldn't the tokenizer just take the current token, check if it's in the list of imported identifiers and color it differently if it is?

**Comment by lierdakil** [comment]
> Couldn't the tokenizer just take the current token, check if it's in the list of imported identifiers and color it differently if it is 
Not unless I hack into Atom's internals and write my own streaming tokenizer for Haskell. Atom's grammar is just a declarative collection of regular expressions, basically. So no real state there.

Besides, collecting identifiers from imports is ***slow*** -- you *don't* want to wait for that to highlight your code. And highlighter is  *synchronous*, by the way, if you were wondering. Also it doesn't retokenize whole buffer if you changed one line (which is good, since large buffer could take seconds to retokenize)

Long story short, what you're describing is entirely orthogonal to actual grammar.

**Comment by LukaHorvat** [comment]
I understand. So there is a possibility of this being implemented as a package, but not as a part of the grammar. Thank you again for your work!

**Comment by lierdakil** [comment]
Let's just say this, when (if) Atom implements multiple text decoration layers, something like what you described will become viable (as a package). Right now, it's purely theoretical, since there's virtually no good way of hacking into highlighter. And highlighting parts of buffer with absolute divs sounds like a nightmare (although [pigments](https://atom.io/packages/pigments) does exactly that). And you're welcome.

**Comment by lierdakil** [comment]
I've played around a bit with Atom's internals, and here's what I came up with: https://atom.io/packages/language-haskell-scoped

Bear in mind that it's pretty hackish and basically of pre-alpha quality.

**Comment by LukaHorvat** [comment]
Man, I don't know what to say. This is some seriously amazing work. It works great!
